### PR TITLE
Add run result package endpoint

### PR DESCRIPTION
**Pull request by dredozubov**
## Summary

Adds a read-only result package endpoint for completed or in-flight runs. The endpoint describes what OD produced without applying anything back to an external source workspace.

## Behavior

- Adds `GET /api/runs/:id/result-package`.
- Adds a matching `od run result-package <runId> [--json]` CLI surface.
- Adds the result-package payload DTO to `packages/contracts`.
- Returns run status/error fields, workspace storage/provenance, event-log location, project summary, and artifact manifest summaries.
- Fails explicitly when OD cannot enumerate the workspace instead of returning a successful empty package.
- Keeps writeback outside OD; this endpoint and CLI command are descriptive only.

## Surface area

- [x] Daemon/API
- [x] Shared contracts
- [x] CLI
- [ ] Web UI
- [ ] Documentation
- [ ] Env vars or config
- [ ] i18n

Web UI is intentionally not checked: this is an embeddability/result-inspection surface for external orchestrators and CLI callers, not a new end-user editing action.

## Merge order

1. https://github.com/nexu-io/open-design/pull/4263
2. https://github.com/nexu-io/open-design/pull/4264
3. https://github.com/nexu-io/open-design/pull/4265

## Validation

- `PATH=/Users/dr/.nvm/versions/node/v24.16.0/bin:$PATH pnpm --filter @open-design/daemon exec vitest run tests/folder-import-route.test.ts tests/runs.test.ts`
- `PATH=/Users/dr/.nvm/versions/node/v24.16.0/bin:$PATH pnpm --filter @open-design/contracts typecheck`
- `PATH=/Users/dr/.nvm/versions/node/v24.16.0/bin:$PATH pnpm --filter @open-design/daemon typecheck`

Refs #3416.

### CODE REVIEW

**Comment by PerishCode** [comment]
@dredozubov I'm holding off on generating review comments for nexu-io/open-design#4265 because this pull request has merge conflicts right now.

Please resolve the conflicts with main and push the updated branch. Once that's done, request or wait for the review to run again and I'll take another look.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @dredozubov — thanks for splitting this into #4263 / #4264 / #4265. The merge-order note in the description is especially helpful here.

One quick PR-body follow-up before pool review picks this up: could you add the **Surface area** checklist and tick the relevant box(es)? The Summary already covers the why / behavior well; the missing piece is just making the affected surface explicit so the next reviewer can scope backend-only vs user-facing impact quickly.

💡 To drive this PR to merge hands-free, paste this to your AI coding agent (Claude Code / Codex / opencode / Cursor …):
`Take over nexu-io/open-design#4265 until it merges — read https://raw.githubusercontent.com/nexu-io/looper/main/skills/pr-takeover/SKILL.md and follow it.`

File: apps/daemon/src/import-export-routes.ts [hunk 1]
```diff
@@ -106,10 +107,12 @@ export function registerImportRoutes(app: Express, ctx: RegisterImportRoutesDeps
       if (!existing) {
         return sendApiError(res, 404, 'PROJECT_NOT_FOUND', 'project not found');
       }
-      const { baseDir } = req.body || {};
+      const { baseDir, orchestratorWorkspace } = req.body || {};
       if (typeof baseDir !== 'string' || !baseDir.trim()) {
         return sendApiError(res, 400, 'BAD_REQUEST', 'baseDir required');
       }
+      const normalizedOrchestratorWorkspace =
+        normalizeOrchestratorWorkspace(orchestratorWorkspace);
```
Comment (nettee) [inline apps/daemon/src/import-export-routes.ts]: When the caller supplies a malformed `orchestratorWorkspace`, `normalizeOrchestratorWorkspace()` collapses it to `null` and the request continues as an ordinary folder import. That silently downgrades an external scratch workspace to local-folder provenance, which changes the downstream writeback semantics this PR is trying to describe. The new field is part of the public import contract now, so treating invalid data as "absent" hides orchestrator bugs instead of surfacing them. Please distinguish "field omitted" from "field present but invalid", return `400` for the latter here and in `/api/import/folder`, and add a regression test that posts an invalid workspace object and asserts rejection.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts [hunk 2]
```diff
@@ -15043,6 +15043,71 @@ export async function startServer({
...
+  app.get('/api/runs/:id/result-package', async (req, res) => {
+    const run = design.runs.get(req.params.id);
+    if (!run) return sendApiError(res, 404, 'NOT_FOUND', 'run not found');
+    const status = design.runs.statusBody(run);
+    const project = run.projectId ? getProject(db, run.projectId) : null;
+    let files: any[] = [];
+    let filesError: string | null = null;
+    if (project) {
+      try {
+        files = await listFiles(PROJECTS_DIR, project.id, { metadata: project.metadata });
```
Comment (nettee) [inline apps/daemon/src/server.ts]: Catching `listFiles()` failures and still returning `200` makes the result package look like a successful empty output when the real problem is that OD could not enumerate the workspace. The changed lines then report `fileCount: 0` and only tuck the failure away in `filesError`, which is exactly the kind of partial-success fallback the repo's fast-fail rules try to avoid. Please surface this as a non-2xx API error (or another explicit failure state that callers must branch on) instead of a normal package, and add a test where the imported folder disappears before `GET /api/runs/:id/result-package`.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts [hunk 3]
```diff
@@ -15043,6 +15043,71 @@ export async function startServer({
     res.json(design.runs.statusBody(run));
   });
 
+  app.get('/api/runs/:id/result-package', async (req, res) => {
```
Comment (nettee) [inline apps/daemon/src/server.ts:15125]: This route introduces a new public payload shape, but it is defined ad hoc in `server.ts` only. `packages/contracts` still stops at `ChatRunStatusResponse`, and `od run` still only exposes `list`/`info`/`cancel`/`watch`, so external callers get a daemon-only surface that can drift from the shared contract. Because the repo requires new user-facing capabilities to land with both the contract type and CLI mirror, please move this response schema into `packages/contracts` and add a matching `od run` subcommand (for example `od run result-package --json`) in the same PR.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found 3 blocking issues in the new import/result-package flow. The inline comments cover input validation, failure semantics, and the missing shared contract/CLI surface for the new endpoint.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/import-export-routes.ts [hunk 4]
```diff
@@ -184,12 +187,18 @@ export function registerImportRoutes(app: Express, ctx: RegisterImportRoutesDeps
 
       const entryFile = await detectEntryFile(normalizedPath);
       const existingMeta = existing.metadata ?? {};
+      const nextOrchestratorWorkspace =
+        normalizedOrchestratorWorkspace ??
```
Comment (nettee) [inline apps/daemon/src/import-export-routes.ts]: At `apps/daemon/src/import-export-routes.ts:190`, `/api/projects/:id/working-dir` falls back to `existingMeta.orchestratorWorkspace` whenever the caller omits `orchestratorWorkspace`. That makes scratch provenance sticky across working-directory swaps: a project imported from an orchestrator workspace once will still report `orchestrator-scratch` after being repointed at an ordinary local folder, so later run status/result-package responses describe the wrong writeback owner. Because this route is the only way to replace the working directory, callers currently have no way to clear that stale metadata. Please treat omission here as "this new working dir is a plain folder-backed project" and drop `orchestratorWorkspace` unless the request explicitly provides a new valid scratch descriptor; add a regression test that imports a scratch workspace, switches to a local folder via `/api/projects/:id/working-dir`, and asserts the next run reports `user-local` provenance.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/import-export-routes.ts (hunk 1 above)
Comment (nettee) [inline apps/daemon/src/import-export-routes.ts]: At `apps/daemon/src/import-export-routes.ts:115`, both import paths feed `orchestratorWorkspace` through `normalizeOrchestratorWorkspace()` and then continue when it returns `null`. That means `{ kind: 'bogus' }`, misspelled fields, or other malformed payloads are treated as if the field was absent, silently downgrading an external scratch import to ordinary local-folder provenance. Because this metadata now drives the result-package contract and writeback interpretation, hiding invalid caller data makes integration bugs look like successful imports. Please distinguish "field omitted" from "field present but invalid", return `400 BAD_REQUEST` for the latter in both `/api/projects/:id/working-dir` and `/api/import/folder`, and add a regression test that posts an invalid workspace object and asserts rejection.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts (hunk 2 above)
Comment (nettee) [inline apps/daemon/src/server.ts]: At `apps/daemon/src/server.ts:15055`, `listFiles()` failures are swallowed into `filesError`, but the endpoint still returns `200` with `project.fileCount: 0` and `artifacts: []`. For folder-backed projects that usually means the workspace disappeared or became unreadable, so callers now see a nominally successful result package that looks like OD produced an empty output instead of surfacing that required enumeration failed. That conflicts with the repo's fast-fail rules and makes downstream orchestration branch on an out-of-band string field instead of HTTP success/failure. Please return a non-2xx API error (or another explicit non-success envelope) when workspace enumeration is required but fails, and add a test that removes the imported folder before `GET /api/runs/:id/result-package`.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts (hunk 3 above)
Comment (nettee) [inline apps/daemon/src/server.ts:15125]: At `apps/daemon/src/server.ts:15046`, this PR adds a new public `result-package` payload and, via `apps/daemon/src/runs.ts`, also adds `workspace` to the existing run status body, but `packages/contracts/src/api/chat.ts` still does not declare either response shape and `apps/daemon/src/cli.ts` still exposes only `od run list|info|cancel|watch`. That leaves daemon HTTP returning fields the shared contract cannot type-check and ships the new capability without the CLI surface that external agents are supposed to drive. Please move the run-status/result-package DTOs into `packages/contracts` and add a matching `od run ... --json` entry point in the same PR so the endpoint, contract, and CLI remain in lockstep.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found 4 blocking issues in the new workspace-provenance/result-package flow. The inline comments cover stale provenance on working-directory swaps, invalid import metadata being accepted silently, partial-success result packages, and the missing shared contract/CLI closure.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @dredozubov — quick follow-up from our side: the current blocking items are already captured well in @nettee's open review on this head, so I won't duplicate them in a second bot review.

The key must-fix set there is: clearing stale scratch provenance on working-directory swaps, rejecting malformed `orchestratorWorkspace` metadata instead of silently downgrading it, failing the result-package route explicitly when workspace enumeration fails, and landing the shared contract / CLI surface alongside the daemon endpoint.

@elihahah666's re-review also lines up with those blockers, with one extra non-blocking note about the currently-unused `isOrchestratorScratchWorkspace` export. Once you've pushed an update for the blocking review, we'll re-check promptly.

File: apps/daemon/src/server.ts [hunk 5]
```diff
@@ -15043,6 +15044,79 @@ export async function startServer({
...
+    const project = run.projectId ? getProject(db, run.projectId) : null;
+    let files: any[] = [];
+    if (project) {
+      try {
+        const projectRoot = resolveProjectDir(PROJECTS_DIR, project.id, project.metadata);
+        const projectRootStat = await fs.promises.stat(projectRoot);
+        if (!projectRootStat.isDirectory()) {
+          throw new Error('workspace root is not a directory');
+        }
+        files = await listFiles(PROJECTS_DIR, project.id, { metadata: project.metadata });
```
Comment (nettee) [inline apps/daemon/src/server.ts]: `result-package` currently mixes two different sources of truth for the same run: `workspace` comes from `design.runs.statusBody(run)` (which snapshots `run.projectMetadata` at run creation), but the file/artifact enumeration here uses the *current* project record (`project.metadata`). If `/api/projects/:id/working-dir` is called after the run starts, this endpoint can report the run's original workspace/provenance while returning files from the replacement folder, so callers get a self-contradictory package. Because this route is explicitly meant for completed or in-flight runs, that race is part of the supported surface. Please derive both halves from the same metadata source: either enumerate from the run snapshot (`run.projectMetadata` / `status.workspace`) or make the whole response explicitly current-project based, and add a regression test that starts a run, swaps the working dir, then asserts the returned workspace and artifacts stay aligned.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found 1 blocking correctness issue in the new result-package flow. The inline comment covers an internal snapshot/current-project mismatch that can make the package describe one workspace while returning artifacts from another.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @dredozubov — one more quick loop-close from us since new reviews landed on this same head.

The current blocking set is now split across two open reviews: @nettee is flagging the snapshot/current-project mismatch in `result-package` (workspace provenance from the run snapshot, artifacts from the current project root), and @elihahah666 is flagging the new od-owned `ENOENT` path that now turns a normal empty project directory into a 500. Those are the two must-fix items to focus on before the next pass.

I’m staying out of the way on duplicate code-review threads while the bot review kill switch is active, but once you push the next revision we’ll re-check promptly.

File: packages/contracts/src/api/projects.ts [hunk 6]
```diff
@@ -130,6 +131,9 @@ export interface ProjectMetadata {
   // it set `baseDir` outside the trusted flow. Privileged: rejected
   // by `POST /api/projects` and `PATCH /api/projects/:id`.
   fromTrustedPicker?: true;
+  // Externally prepared scratch workspace provenance. OD may read/write
+  // metadata.baseDir, but source authority and writeback stay outside OD.
+  orchestratorWorkspace?: OrchestratorWorkspace;
```
Comment (nettee) [inline packages/contracts/src/api/projects.ts:136]: Adding `orchestratorWorkspace` to generic `ProjectMetadata` makes it writable through the existing `PATCH /api/projects/:id` API, but this PR only validates the field on `/api/import/folder` and `/api/projects/:id/working-dir`. That leaves a live public entrypoint where malformed provenance is still accepted and later normalized away: for example, a PATCH that sends `{ metadata: { ..., orchestratorWorkspace: { kind: 'scratch', source_reference: 'typo' } } }` succeeds today, then `projectWorkspaceProvenance()` drops the bad object and subsequent run status / result-package responses silently downgrade to `user-local` or `null`. This reintroduces the same silent-fallback bug the import-route changes were meant to eliminate, just through a different contract surface. Please either validate and reject `orchestratorWorkspace` in the generic project patch route, or keep it out of patchable `ProjectMetadata` entirely, and add a regression test for the PATCH path.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found 1 blocking API-consistency issue in the new orchestrator-workspace metadata flow. The inline comment covers a remaining public entrypoint that still accepts silent provenance downgrades.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @dredozubov — quick status sync on the new head.

The blocking set is narrower now: @nettee's current review is down to one must-fix item on API consistency — `orchestratorWorkspace` is now part of patchable project metadata, so the generic `PATCH /api/projects/:id` path also needs to reject malformed provenance instead of silently accepting and later normalizing it away.

@elihahah666's latest pass is non-blocking on this head: the remaining notes there are the sandbox allowlist bypass concern for self-asserted scratch workspaces and dropping the module-level `// @ts-nocheck`. Once the PATCH-path validation issue is addressed and pushed, we'll re-check promptly.

File: packages/contracts/src/api/projects.ts (hunk 6 above)
Comment (dredozubov) [reply packages/contracts/src/api/projects.ts:136]: Addressed in `b0c25f7f8`. Generic project create/patch now validates `metadata.orchestratorWorkspace` through the same parser used by `/api/import/folder` and `/api/projects/:id/working-dir`, so malformed fields like `source_reference` are rejected instead of being stored and later normalized away. Added PATCH regression coverage in `apps/daemon/tests/projects-routes.test.ts`.

File: apps/daemon/src/import-export-routes.ts [hunk 7]
```diff
@@ -177,19 +189,26 @@ export function registerImportRoutes(app: Express, ctx: RegisterImportRoutesDeps
       ) {
         return sendApiError(res, 400, 'BAD_REQUEST', 'cannot point at the data directory');
       }
-      const sandboxReason = sandboxImportedProjectRootUnavailableReason(normalizedPath);
+      const sandboxReason = normalizedOrchestratorWorkspace
+        ? null
+        : sandboxImportedProjectRootUnavailableReason(normalizedPath);
```
[the remark this answers is not in the document]
Comment (dredozubov) [reply apps/daemon/src/import-export-routes.ts]: Addressed in `b0c25f7f8`. The sandbox import-root bypass now requires both normalized scratch provenance and a trusted desktop import token (`trustedPickerImport`). A self-asserted `orchestratorWorkspace: { kind: "scratch" }` no longer skips `OD_SANDBOX_IMPORT_ALLOWED_ROOTS`. Added regression coverage for the unauthenticated rejection path and the trusted-token path.

File: apps/daemon/src/workspace-contract.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
+// @ts-nocheck
```
[the remark this answers is not in the document]
Comment (dredozubov) [reply apps/daemon/src/workspace-contract.ts]: Addressed in `b0c25f7f8`. Removed `// @ts-nocheck` from `workspace-contract.ts`; daemon typecheck now covers the module. That exposed the optional `writeback` spread into required `RunWorkspaceProvenance.writeback`, so the result-package provenance now sets `writeback: "external"` explicitly.

File: apps/daemon/src/project-routes.ts [hunk 9]
```diff
@@ -1085,6 +1086,19 @@ export function registerProjectRoutes(app: Express, ctx: RegisterProjectRoutesDe
             'fromTrustedPicker can only be set via POST /api/import/folder',
           );
         }
+        if ('orchestratorWorkspace' in metadata) {
```
Comment (nettee) [inline apps/daemon/src/project-routes.ts:1096]: This still allows callers to create an impossible public state: `POST /api/projects` rejects client-supplied `baseDir`, but it now accepts `metadata.orchestratorWorkspace` as long as the shape parses. From there, `projectWorkspaceProvenance()` in `apps/daemon/src/workspace-contract.ts` reports `storage.kind: "folder-backed"` whenever `orchestratorWorkspace` is present, even if `baseDir` is missing, while `resolveProjectDir()`/`listFiles()` continue to use `.od/projects/<id>` because there is no absolute external root. The result is that `/api/runs/:id` and `/api/runs/:id/result-package` can describe an OD-owned project as an orchestrator scratch workspace. Please reject `orchestratorWorkspace` on the generic create/patch routes unless the request is also establishing or preserving a real folder-backed `baseDir` via the import/working-dir flow.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found 2 blocking correctness issues in the new orchestrator-workspace metadata flow. The inline comments cover impossible scratch provenance on generic project writes and provenance being dropped on ordinary metadata patches.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

Location: apps/daemon/src/project-routes.ts RIGHT line 1386

`updateProject()` replaces `metadata` wholesale, but this restamp preserves only `baseDir`/import markers. On an imported scratch project, a normal PATCH like `{ metadata: { kind: "prototype" } }` will silently drop `existingMeta.orchestratorWorkspace`, and the next run flips from `orchestrator-scratch`/`external` to `user-local`/`in-place` because `projectWorkspaceProvenance()` falls back to the bare-`baseDir` branch. That corrupts the new workspace/result-package surface even though the caller never asked to change provenance. Please preserve `existingMeta.orchestratorWorkspace` here unless the caller is going through `/api/projects/:id/working-dir` to replace or clear it, and add a regression test that patches unrelated metadata on a scratch project and verifies the run workspace provenance stays `orchestrator-scratch`.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

Inline comment could not be anchored: inline anchor is outside the PR diff anchorable ranges

**Comment by lefarcen** [comment]
Hey @dredozubov — quick status sync on the current head.

The review picture is cleaner now: @elihahah666 has approved this revision, and the remaining blocking set is down to the two correctness items in @nettee's current review. The must-fix points there are (1) preserving `orchestratorWorkspace` on ordinary metadata patches unless the caller is explicitly replacing/clearing working-dir provenance, and (2) rejecting impossible generic create/patch states where `orchestratorWorkspace` is set without a real folder-backed `baseDir`.

I’m avoiding duplicate code-review threads while the bot review kill switch is active, but once those two items are addressed and pushed we’ll re-check promptly.

File: apps/daemon/src/project-routes.ts (hunk 9 above)
Comment (dredozubov) [reply apps/daemon/src/project-routes.ts:1096]: Addressed in `9214e6a3e`. Generic `POST /api/projects` now rejects `metadata.orchestratorWorkspace` because that route cannot establish a folder-backed `baseDir`. Generic `PATCH` now only allows `orchestratorWorkspace` when preserving the already-imported folder-backed provenance; attempts to add or change it through the generic route are rejected, and ordinary metadata patches on scratch imported projects re-stamp the existing provenance.

Added regressions for POST smuggling, OD-owned PATCH smuggling, and unrelated metadata PATCH preserving scratch provenance.

Validation:
- `pnpm --filter @open-design/daemon exec vitest run tests/folder-import-route.test.ts tests/projects-routes.test.ts`
- `pnpm --filter @open-design/daemon typecheck`

File: apps/daemon/src/project-routes.ts [hunk 10]
```diff
@@ -1371,6 +1412,9 @@ export function registerProjectRoutes(app: Express, ctx: RegisterProjectRoutesDe
             ...(existingMeta.fromTrustedPicker === true
               ? { fromTrustedPicker: true as const }
               : {}),
+            ...(existingMeta.orchestratorWorkspace
```
Comment (nettee) [inline apps/daemon/src/project-routes.ts:1426]: Preserving `existingMeta.orchestratorWorkspace` here fixes the ordinary object-patch case, but `UpdateProjectRequest` still allows `metadata: null` (`packages/contracts/src/api/projects.ts`). That value skips the entire `typeof patch.metadata === 'object'` guard above, and `updateProject()` then writes `metadata_json = NULL` wholesale. On an imported project this drops `baseDir`, `fromTrustedPicker`, and scratch provenance in one request, so later reads/writes silently fall back to `.od/projects/<id>` and the new run/result-package surfaces describe the wrong workspace. Please either reject `metadata: null` for imported projects or normalize it to a restamped object that preserves the immutable import fields, and add a regression test for the null-patch path.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found 1 blocking correctness issue in the current metadata-patch flow. The inline comment covers an imported-project detach path that still bypasses the new provenance preservation logic.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @dredozubov — thanks for the quick follow-up here.

The blocking set on this head now looks down to the one must-fix path in @nettee's current review: `PATCH /api/projects/:id` can still clear imported-project provenance wholesale when callers send `metadata: null`, so that detach/null-patch path needs to reject or restamp instead of dropping `baseDir` / scratch provenance.

Everything else on this revision looks like it moved forward cleanly. Once that null-patch case is covered and pushed, we'll re-check promptly.

**Comment by PerishCode** [comment]
@dredozubov I'm holding off on generating review comments for nexu-io/open-design#4265 because this pull request has merge conflicts right now.

Please resolve the conflicts with main and push the updated branch. Once that's done, request or wait for the review to run again and I'll take another look.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/project-routes.ts (hunk 10 above)
Comment (dredozubov) [reply apps/daemon/src/project-routes.ts:1426]: Addressed in `448592310`. The generic project PATCH route now rejects `metadata: null` when the existing project has imported/folder-backed metadata, so a null patch cannot clear `baseDir`, `fromTrustedPicker`, project-location fields, or scratch provenance and detach the project back to `.od/projects/<id>`.

I also merged current `origin/main` into the branch to clear the reported conflicts and kept the result-package route ahead of the generic `/api/runs/:id` route so the endpoint keeps returning the package envelope.

Added a regression for the null-patch path on a project-location project.

Validation:
- `pnpm --filter @open-design/contracts build`
- `pnpm --filter @open-design/daemon exec vitest run tests/folder-import-route.test.ts tests/projects-routes.test.ts`
- `pnpm --filter @open-design/daemon typecheck`

**Review by nettee — APPROVED** [review]
@dredozubov I re-reviewed the final changed ranges at `448592310e`: the result-package route now stays aligned with the run's workspace snapshot, the generic project create/patch paths no longer let callers smuggle or accidentally drop orchestrator provenance, and the CLI/contracts closure for `od run result-package` is in place. The regression matrix added here covers the cases that were risky in earlier rounds: malformed provenance input, sandbox scratch imports, missing folder-backed workspaces, empty OD-owned runs, working-dir swaps, and imported-project metadata patch preservation.

This looks ready to merge. Nice work closing the loop on the earlier review findings without widening the surface.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Hey @dredozubov — great persistence working through six rounds of feedback here. The current head (44859231) is looking clean:

- @nettee approved ✅
- @elihahah666 approved ✅
- CI all green ✅
- 0 unresolved review threads ✅

One stack note: #4264 (Expose run workspace provenance) is still open — it's #2 in the merge order, so that one needs to land before this PR. Maintainer has been looped in on both. 🎉

**Comment by kokisanai** [comment]
Hi @dredozubov!

This PR looks quiet for a bit, so I'm checking in to make sure it doesn't get stuck.

If you are waiting on review, feedback, or a specific unblock from the team, feel free to say so here. If faster coordination helps, you can also drop into our contributors channel in Discord:
https://discord.gg/3C6EWXbdQQ

We want to help move this forward instead of letting it stall.

**Comment by kokisanai** [comment]
\nHi @dredozubov!\n\nThis PR looks quiet for a bit, so I'm checking in to make sure it doesn't get stuck.\n\nIf you are waiting on review, feedback, or a specific unblock from the team, feel free to say so here. If faster coordination helps, you can also drop into our contributors channel in Discord:\nhttps://discord.gg/3C6EWXbdQQ\n\nWe want to help move this forward instead of letting it stall.

**Comment by lefarcen** [comment]
@dredozubov — just to add context to the check-in above: nothing's stuck on your end. @nettee approved the current head, CI is all green, and there are no unresolved threads.

The PR is quiet because of the stack dependency — #4264 needs to land first since this is PR 3/3 in the merge order you called out. Once #4264 merges, this one is good to go. No action needed from you.

**Comment by kokisanai** [comment]
Hi @dredozubov!

You are now in the 10-19 merged PR range in Open Design, which means you're getting very close to the next contributor threshold. Thanks for the steady contributions here.

If you want to keep the momentum going, feel free to pick up another scoped issue and comment /claim on it. If you'd like, we can also help route you to a good next task based on the areas you've already been contributing in.

We also have a contributors channel in Discord if you want faster coordination with the team and other contributors:
https://discord.gg/3C6EWXbdQQ

Thanks again for continuing to build with us.

**Comment by lefarcen** [comment]
Hey @dredozubov — and congrats on getting this one over the line.

The workspace-provenance / result-package stack landed cleanly, and thanks again for sticking with the full #4263 / #4264 / #4265 split until it merged. If you want another scoped follow-up in this area, happy to help point you at one.
